File: Programmers/LV2/12981.py
# 끝말잇기
def solution(n, words):
    answer = [0,0]  # 답 초기값
    gameState = []  # 게임 진행 리스트
    # 게임이 진행 완료까지
    while words:
        # 현재 라운드의 값
        now = words.pop(0)
        # 만약 진행 리스트에 중복된 값이 없을떄 -> 예외처리1 중복 x
        if now not in gameState:
            # 게임 진행리스트가 존재할때 현재 라운드의값이랑 끝말잇기가 되었는지 확인 -> 예외처리2 끝말잇기 체크
            if len(gameState)> 0 and gameState[-1][-1] != now[0]:
                answer = [(len(gameState) % n) +1, ((len(gameState) // n)) +1]
                break
            gameState.append(now)
        # 진행리스트에 중복값이 존재할때 -> 예외처리 1
        else:
            answer = [(len(gameState) % n) +1, ((len(gameState) // n)) +1]
            break
    print(answer)
    return answer

File: Programmers/LV2/12981.js
// 영어 끝말잇기
// 한바퀴돌면 다시 첫사람으로
// 이전에 등장한 단어 사용 금지 (큐나 스택에서 빼야할듯)
// 한글자 안됨.
// 시간 복잡도 O(N)
function solution(n, words) {
  let answer = [0, 0];
  let status = new Set(); // 사용된 단어를 저장할 set
  let preWord = words[0]; // 이전 단어
  status.add(preWord);
  // 끝말잇기 시작
  for (let i = 1; i < words.length; i++) {
    let curWord = words[i]; // 현재 단어
    // 끝말 잇기가 성공했을때
    if (preWord[preWord.length - 1] === curWord[0]) {
      preWord = curWord; // 이전단어 업데이트
      // 만약 이전에 사용했다면 break
      if (!status.has(curWord)) status.add(curWord);
      else {
        answer = i + 1;
        break;
      }
    } else { // 끝말잇기가 틀렸을 때
      answer = i + 1;
      break;
    }
  }
  // 플레이어와 순서 찾기
  let playerNUm = (answer % n) - 1 < 0 ? n : answer % n;
  let seq = Math.ceil(answer / n);
  if (playerNUm && seq) answer = [playerNUm, seq];
  return answer;
}
